### DISCUSSION TITLE
Start userscript from outside a Qutebrowser instance

### DISCUSSION THREAD

**Discussion by wgmayer0** [qutebrowser/qutebrowser#6960] [Q&A]
Can a userscript be ran from a bash script without actually hitting a keyboard shortcut in Qutebrowser? I get my input for searches using vim and then like to send that query to Qutebrowser. I use xbindkeys to bind `alt + space` to launch vim so I can type my search query. It takes Qutebrowser 1-2 seconds to boot up so in the interest of speed, I would like Qutebrowser to boot in parallel while I am typing my query. Once I hit `ZZ` on the keyboard to save my vim search query, I have xdotool pressing `o` to start the userscript. I would like a more elegant solution where I don't have to simulate keystrokes and can tell Qutebrowser directly what to search for.

Userscript:
```
#!/bin/bash

search_file=/home/wgm/.search
if [ ! -f "$search_file" ]
then
	touch $search_file
	xterm -e vim $search_file
fi
QUERY=$(< $search_file)
rm $search_file
if [ ! -z "$QUERY" ] 
then
    echo "open "$QUERY >> "$QUTE_FIFO"
fi

```

alt + space bash script
```
#!/bin/bash

# start qutebrowser and track its PID
qutebrowser &
pid=$!

# define the search file location
search_file=/home/wgm/.search

# create the file that holds the search term
touch $search_file

# create an xterm window to input the user's search term
sleep 1.5s
xterm -e vim $search_file
sleep 0.4s

# activate qutebrowser and press o so that the search is ran
xdotool windowactivate $pid
xdotool key o

```

**Comment by mschilli87** [comment]
I didn't test this but I would assume a simple `qutebrowser ':spawn --userscript'` instead of your entire `xdotool` part to work. Basically launch qutebrowser running whatever command you have bound to `o`. If a browser instance is running already, it should be used automatically. Otherwise, a new instance will be spawned.

**Comment by wgmayer0** [comment]
So that worked. I was able to use `qutebrowser ':spawn --userscript
$HOME/.local/share/qutebrowser/userscripts/{name of script}'` to
send commands to Qutebrowser but if I have multiple Qutebrowser windows open, how do I configure the script to send it to the most recent window. I am not sure how the program determines which window to send it to.

**Reply by mschilli87** [reply]
I think you are looking for the `--target` parameter of qutebrowser. The default is `auto` and IIRC it will pick the qutebrowser window that last had focus. But I won't be of much more help here since I myself tend to use a single window only. :slightly_smiling_face:

**Reply by wgmayer0** [reply]
How hard would a `--target` parameter be to write? @The-Compiler

**Reply by The-Compiler** [reply]
`--target` already exists, so I assume you mean extending it (or adding something new) to select a window from the commandline? Making it take window IDs or so probably wouldn't be hard, but I'm not sure if/how that solves your issue: How are you going to determine the window ID of the window you want to send it to?

**Reply by wgmayer0** [reply]
Here is a script bound to my window manager:
```
#!/bin/bash

qutebrowser &
search_file=/home/wgm/.search
touch $search_file
xterm -e vim $search_file
qutebrowser ':spawn --userscript /home/wgm/.local/share/qutebrowser/userscripts/new_search'

```

For context, here is the userscript it refers to:
```
#!/bin/bash

search_file=/home/wgm/.search
if [ ! -f "$search_file" ]
then
	touch $search_file
	xterm -e vim $search_file
fi
QUERY=$(< $search_file)
rm $search_file
if [ ! -z "$QUERY" ] 
then
    echo "open "$QUERY >> "$QUTE_FIFO"

```

I have 2 problems:

1. I cannot apply my query to line 3 of the search script, `qutebrowser &`
2. Qutebrowser steals focus from `.search` xterm window (maybe outside scope of this thread. `xmonad` thread maybe?) https://github.com/xmonad/xmonad

**Reply by The-Compiler** [reply]
I still don't follow. Why does your script need to invoke qutebrowser twice?

**Reply by wgmayer0** [reply]
That way qutebrowser starts loading as I am typing the query. Also, if I don't do that, qutebrowser is still not ready yet and I get this:
![1643131606](https://user-images.githubusercontent.com/73801421/151028039-a435176a-8e17-42c0-b7ec-cfff9dc715a0.png)

I then have to hit `o` again for it to work.

**Comment by wgmayer0** [comment]
Thanks yeah I don't use tabs. I have xmonad manage the windows

**Reply by mschilli87** [reply]
I guess you could focus the window of your choice via xdotool and then use the spawn command as suggested.
Still an ugly hack but maybe less ugly than before until somebody comes by with a way to select the target window from the CLI. :wink:

**Reply by wgmayer0** [reply]
You're right - this is less ugly than before for sure.

**Reply by wgmayer0** [reply]
```
$ qutebrowser "open msn.com" & echo $! && qutebrowser "open google.com" & echo $!

[1] 1390613
[2] 1390614

14:53:25 INFO: Opening in existing instance

$ xdotool windowactivate 1390613

Your windowmanager claims not to support _NET_ACTIVE_WINDOW, so the attempt to activate the window was aborted.
xdo_activate_window on window:1390613 reported an error

[2]+  Done                    echo $! && qutebrowser "open google.com"
14:54:46 ERROR: Load error: HTTP ERROR 403
14:54:48 ERROR: Load error: HTTP ERROR 403
```

This was working before on `xmonad` window manager, but now I am not able to use it.

**Reply by wgmayer0** [reply]
Any ideas @The-Compiler ?

**Reply by The-Compiler** [reply]
Not really, no. #6369 has some notes on this.
